### PR TITLE
Leds

### PR DESCRIPTION
**Pull request by Lewis-Seiden**

### CODE REVIEW

File: src/main/java/frc/robot/subsystems/leds/LEDIOSim.java [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+// the WPILib BSD license file in the root directory of this project.
+
+package frc.robot.subsystems.leds;
+
+import edu.wpi.first.wpilibj.AddressableLED;
+import edu.wpi.first.wpilibj.AddressableLEDBuffer;
+import edu.wpi.first.wpilibj.util.Color;
+
+// This is a copy of LEDIOReal right now, but if we move leds to separate ports sim breaks w
+// ledioreal
```
Comment (Awesomeplayer165) [inline src/main/java/frc/robot/subsystems/leds/LEDIOSim.java:12]:
```suggestion
// LEDIOReal
```

File: src/main/java/frc/robot/subsystems/leds/LEDIOSim.java [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+public class LEDIOSim implements LEDIO {
+  AddressableLED led;
+  AddressableLEDBuffer buffer;
+
+  public LEDIOSim() {
+    led = new AddressableLED(3);
+    buffer = new AddressableLEDBuffer(LEDSubsystem.LED_LENGTH);
+    led.setLength(buffer.getLength());
```
Comment (Awesomeplayer165) [inline src/main/java/frc/robot/subsystems/leds/LEDIOSim.java:20]: Line break

File: src/main/java/frc/robot/subsystems/leds/LEDSubsystem.java [hunk 3]
```diff
@@ -0,0 +1,124 @@
...
+
+public class LEDSubsystem extends SubsystemBase {
+  public static final int LED_LENGTH = 36;
+  LEDIO io;
+  LEDIOInputsAutoLogged inputs = new LEDIOInputsAutoLogged();
+
+  Color[] colors = new Color[LED_LENGTH];
+
+  double rainbowStart = 0;
+  double dashStart = 0;
```
Comment (Awesomeplayer165) [inline src/main/java/frc/robot/subsystems/leds/LEDSubsystem.java:27]: Should any of these be `private`?

File: src/main/java/frc/robot/subsystems/leds/LEDSubsystem.java [hunk 4]
```diff
@@ -0,0 +1,124 @@
...
+          for (int i = (int) dashStart; i < dashStart + dashLength; i++) {
+            setIndex(i % LED_LENGTH, colorDash.get());
+          }
+
+          dashStart += LED_LENGTH * frequency * 0.020;
+          dashStart %= LED_LENGTH;
+        });
+  }
+
+  public Command defaultStateDisplay(BooleanSupplier enabled, BooleanSupplier targetIsSpeaker) {
```
Comment (Awesomeplayer165) [inline src/main/java/frc/robot/subsystems/leds/LEDSubsystem.java:96]: Postfix Cmd?

File: src/main/java/frc/robot/subsystems/leds/LEDIOSim.java [hunk 5]
```diff
@@ -0,0 +1,34 @@
...
+
+import edu.wpi.first.wpilibj.AddressableLED;
+import edu.wpi.first.wpilibj.AddressableLEDBuffer;
+import edu.wpi.first.wpilibj.util.Color;
+
+// This is a copy of LEDIOReal right now, but if we move leds to separate ports sim breaks w
+// ledioreal
+public class LEDIOSim implements LEDIO {
+  AddressableLED led;
+  AddressableLEDBuffer buffer;
```
Comment (Awesomeplayer165) [inline src/main/java/frc/robot/subsystems/leds/LEDIOSim.java:15]: Private or public?

File: src/main/java/frc/robot/subsystems/leds/LEDIOReal.java [hunk 6]
```diff
@@ -0,0 +1,32 @@
...
+
+package frc.robot.subsystems.leds;
+
+import edu.wpi.first.wpilibj.AddressableLED;
+import edu.wpi.first.wpilibj.AddressableLEDBuffer;
+import edu.wpi.first.wpilibj.util.Color;
+
+public class LEDIOReal implements LEDIO {
+  AddressableLED led;
+  AddressableLEDBuffer buffer;
```
Comment (Awesomeplayer165) [inline src/main/java/frc/robot/subsystems/leds/LEDIOReal.java:13]: Public or private?

**Review by Awesomeplayer165 — CHANGES_REQUESTED** [review]
General consistency stuff.

File: src/main/java/frc/robot/subsystems/leds/LEDIOSim.java (hunk 2 above)
Comment (Lewis-Seiden) [reply src/main/java/frc/robot/subsystems/leds/LEDIOSim.java:20]: ?

File: src/main/java/frc/robot/subsystems/leds/LEDIOSim.java [hunk 7]
```diff
@@ -0,0 +1,34 @@
...
+// This is a copy of LEDIOReal right now, but if we move leds to separate ports sim breaks w
+// LEDIOReal
+public class LEDIOSim implements LEDIO {
+  private final AddressableLED led;
+  private final AddressableLEDBuffer buffer;
+
+  public LEDIOSim() {
+    led = new AddressableLED(3);
+    buffer = new AddressableLEDBuffer(LEDSubsystem.LED_LENGTH);
+    led.setLength(buffer.getLength());
```
Comment (Awesomeplayer165) [inline src/main/java/frc/robot/subsystems/leds/LEDIOSim.java:20]: group similar ideas together for consistency with rest of codebase.

```suggestion
    buffer = new AddressableLEDBuffer(LEDSubsystem.LED_LENGTH);
    
    led.setLength(buffer.getLength());
```

File: src/main/java/frc/robot/subsystems/leds/LEDSubsystem.java [hunk 8]
```diff
@@ -0,0 +1,123 @@
...
+
+  /** Sets the first portion of the leds to a color, and the rest off */
+  public Command setProgressCmd(Color color, DoubleSupplier progress) {
+    return this.run(
+        () -> {
+          for (int i = 0; i < LED_LENGTH; i++) {
+            if (i < progress.getAsDouble() * LED_LENGTH) {
+              setIndex(i, color);
+            } else {
+              setIndex(i, Color.kBlack);
```
Comment (Awesomeplayer165) [inline src/main/java/frc/robot/subsystems/leds/LEDSubsystem.java:65]:
```suggestion
              setIndex(i, i < progress.getAsDouble() * LED_LENGTH ? color : Color.kBlack);
```

File: src/main/java/frc/robot/subsystems/leds/LEDSubsystem.java [hunk 9]
```diff
@@ -0,0 +1,123 @@
...
+import org.littletonrobotics.junction.Logger;
+
+public class LEDSubsystem extends SubsystemBase {
+  public static final int LED_LENGTH = 36;
+
+  private final LEDIO io;
+  private final LEDIOInputsAutoLogged inputs = new LEDIOInputsAutoLogged();
+
+  private double rainbowStart = 0;
+  private double dashStart = 0;
```
Comment (Awesomeplayer165) [inline src/main/java/frc/robot/subsystems/leds/LEDSubsystem.java:26]: add comments as to what time units or meaning start has?

**Review by Awesomeplayer165 — CHANGES_REQUESTED** [review]
general small things I didn't see earlier.

**Review by Awesomeplayer165 — APPROVED** [review]

**Pull request merged**